File: logcli/filters.py
"""Filtering logic for log entries."""

import re
import ipaddress
from datetime import datetime
from typing import Dict, Any, List, Optional, Set, Tuple
from .config import DEFAULT_FILTERS


class LogFilter:
    """Filters log entries based on various criteria."""
    
    def __init__(self):
        self.filters = DEFAULT_FILTERS.copy()
        self._compiled_path_patterns = []
        self._ip_networks = []
        self._update_compiled_patterns()
    
    def update_filters(self, **kwargs) -> None:
        """Update filter settings."""
        self.filters.update(kwargs)
        self._update_compiled_patterns()
    
    def _update_compiled_patterns(self) -> None:
        """Compile regex patterns and IP networks for performance."""
        # Compile path patterns
        self._compiled_path_patterns = []
        for pattern in self.filters.get('paths', []):
            try:
                self._compiled_path_patterns.append(re.compile(pattern, re.IGNORECASE))
            except re.error:
                # Skip invalid regex patterns
                pass
        
        # Parse IP networks
        self._ip_networks = []
        for ip_range in self.filters.get('ip_ranges', []):
            try:
                self._ip_networks.append(ipaddress.ip_network(ip_range, strict=False))
            except ValueError:
                # Skip invalid IP ranges
                pass
    
    def should_include(self, log_entry: Dict[str, Any]) -> bool:
        """Check if a log entry should be included based on current filters."""
        
        # Country filter
        if self.filters.get('countries'):
            country = log_entry.get('country', '').upper()
            if country not in [c.upper() for c in self.filters['countries']]:
                return False
        
        # Status code filter
        if self.filters.get('status_codes'):
            status = log_entry.get('status', 0)
            if status not in self.filters['status_codes']:
                return False
        
        # Bot filter
        if self.filters.get('exclude_bots', False):
            if log_entry.get('is_bot', False):
                return False
        
        # IP range filter
        if self._ip_networks:
            ip = log_entry.get('ip')
            if ip and not any(ip in network for network in self._ip_networks):
                return False
        
        # Time range filter
        time_range = self.filters.get('time_range')
        if time_range:
            timestamp = log_entry.get('timestamp')
            if timestamp:
                start_time, end_time = time_range
                
                # Handle timezone-aware vs naive datetime comparison
                if timestamp.tzinfo is not None and start_time and start_time.tzinfo is None:
                    timestamp = timestamp.replace(tzinfo=None)
                elif timestamp.tzinfo is None and start_time and start_time.tzinfo is not None:
                    start_time = start_time.replace(tzinfo=None)
                if timestamp.tzinfo is not None and end_time and end_time.tzinfo is None:
                    timestamp = timestamp.replace(tzinfo=None)
                elif timestamp.tzinfo is None and end_time and end_time.tzinfo is not None:
                    end_time = end_time.replace(tzinfo=None)
                
                # Apply time range filter
                if start_time and end_time:
                    if not (start_time <= timestamp <= end_time):
                        return False
                elif start_time:
                    if timestamp < start_time:
                        return False
                elif end_time:
                    if timestamp > end_time:
                        return False
        
        # HTTP method filter
        if self.filters.get('methods'):
            method = log_entry.get('method', '').upper()
            if method not in [m.upper() for m in self.filters['methods']]:
                return False
        
        # Path pattern filter
        if self._compiled_path_patterns:
            path = log_entry.get('path', '')
            if not any(pattern.search(path) for pattern in self._compiled_path_patterns):
                return False
        
        return True
    
    def get_active_filters(self) -> Dict[str, Any]:
        """Get currently active filters (non-empty ones)."""
        active = {}
        for key, value in self.filters.items():
            if value:  # Non-empty list, non-None, non-False
                active[key] = value
        return active
    
    def clear_filters(self) -> None:
        """Clear all filters."""
        self.filters = DEFAULT_FILTERS.copy()
        self._update_compiled_patterns()
    
    def add_country_filter(self, countries: List[str]) -> None:
        """Add countries to filter."""
        current = set(self.filters.get('countries', []))
        current.update(countries)
        self.filters['countries'] = list(current)
    
    def remove_country_filter(self, countries: List[str]) -> None:
        """Remove countries from filter."""
        current = set(self.filters.get('countries', []))
        current.difference_update(countries)
        self.filters['countries'] = list(current)
    
    def add_status_filter(self, status_codes: List[int]) -> None:
        """Add status codes to filter."""
        current = set(self.filters.get('status_codes', []))
        current.update(status_codes)
        self.filters['status_codes'] = list(current)
    
    def remove_status_filter(self, status_codes: List[int]) -> None:
        """Remove status codes from filter."""
        current = set(self.filters.get('status_codes', []))
        current.difference_update(status_codes)
        self.filters['status_codes'] = list(current)
    
    def toggle_bot_filter(self) -> None:
        """Toggle bot exclusion filter."""
        self.filters['exclude_bots'] = not self.filters.get('exclude_bots', False)
    
    def set_time_range(self, start_time: Optional[datetime], end_time: Optional[datetime]) -> None:
        """Set time range filter."""
        if start_time and end_time:
            self.filters['time_range'] = (start_time, end_time)
        else:
            self.filters['time_range'] = None
    
    def add_ip_range(self, ip_range: str) -> None:
        """Add IP range to filter."""
        current = self.filters.get('ip_ranges', [])
        if ip_range not in current:
            current.append(ip_range)
            self.filters['ip_ranges'] = current
            self._update_compiled_patterns()
    
    def remove_ip_range(self, ip_range: str) -> None:
        """Remove IP range from filter."""
        current = self.filters.get('ip_ranges', [])
        if ip_range in current:
            current.remove(ip_range)
            self.filters['ip_ranges'] = current
            self._update_compiled_patterns()
    
    def add_path_pattern(self, pattern: str) -> None:
        """Add path pattern to filter."""
        current = self.filters.get('paths', [])
        if pattern not in current:
            current.append(pattern)
            self.filters['paths'] = current
            self._update_compiled_patterns()
    
    def remove_path_pattern(self, pattern: str) -> None:
        """Remove path pattern from filter."""
        current = self.filters.get('paths', [])
        if pattern in current:
            current.remove(pattern)
            self.filters['paths'] = current
            self._update_compiled_patterns()


class FilterPresets:
    """Predefined filter presets for common use cases."""
    
    @staticmethod
    def errors_only() -> Dict[str, Any]:
        """Filter to show only error responses (4xx, 5xx)."""
        return {
            'status_codes': list(range(400, 600)),
            'countries': [],
            'exclude_bots': False,
            'ip_ranges': [],
            'time_range': None,
            'methods': [],
            'paths': []
        }
    
    @staticmethod
    def success_only() -> Dict[str, Any]:
        """Filter to show only successful responses (2xx)."""
        return {
            'status_codes': list(range(200, 300)),
            'countries': [],
            'exclude_bots': False,
            'ip_ranges': [],
            'time_range': None,
            'methods': [],
            'paths': []
        }
    
    @staticmethod
    def no_bots() -> Dict[str, Any]:
        """Filter to exclude all bot traffic."""
        return {
            'status_codes': [],
            'countries': [],
            'exclude_bots': True,
            'ip_ranges': [],
            'time_range': None,
            'methods': [],
            'paths': []
        }
    
    @staticmethod
    def api_only() -> Dict[str, Any]:
        """Filter to show only API endpoints."""
        return {
            'status_codes': [],
            'countries': [],
            'exclude_bots': False,
            'ip_ranges': [],
            'time_range': None,
            'methods': [],
            'paths': [r'^/api/.*']
        }
    
    @staticmethod
    def high_response_time(threshold: float = 1.0) -> Dict[str, Any]:
        """Filter for requests with high response times."""
        # Note: This would need to be implemented in the aggregator
        # as it requires access to response_time values
        return DEFAULT_FILTERS.copy()
    
    @staticmethod
    def recent_activity(hours: int = 1) -> Dict[str, Any]:
        """Filter for recent activity within specified hours."""
        from datetime import datetime, timedelta
        end_time = datetime.now()
        start_time = end_time - timedelta(hours=hours)
        
        return {
            'status_codes': [],
            'countries': [],
            'exclude_bots': False,
            'ip_ranges': [],
            'time_range': (start_time, end_time),
            'methods': [],
            'paths': []
        }


def create_filter_from_preset(preset_name: str, **kwargs) -> LogFilter:
    """Create a LogFilter instance from a preset."""
    log_filter = LogFilter()
    
    preset_methods = {
        'errors_only': FilterPresets.errors_only,
        'success_only': FilterPresets.success_only,
        'no_bots': FilterPresets.no_bots,
        'api_only': FilterPresets.api_only,
        'recent_activity': FilterPresets.recent_activity,
    }
    
    if preset_name in preset_methods:
        if preset_name == 'recent_activity' and 'hours' in kwargs:
            filters = preset_methods[preset_name](kwargs['hours'])
        else:
            filters = preset_methods[preset_name]()
        
        log_filter.update_filters(**filters)
    
    return log_filter

File: internal/filters/filters.go
package filters

import (
	"net"
	"regexp"
	"strings"
	"time"

	"github.com/hpowernl/hlogcli/pkg/models"
)

// LogFilter provides filtering functionality for log entries
type LogFilter struct {
	countries       map[string]bool
	statusCodes     map[int]bool
	excludeBots     bool
	ipRanges        []*net.IPNet
	timeRange       *models.TimeRange
	methods         map[string]bool
	pathPatterns    []*regexp.Regexp
	ignoreBlocked   bool
	blockedStatuses map[int]bool
}

// NewLogFilter creates a new log filter with default settings
func NewLogFilter() *LogFilter {
	return &LogFilter{
		countries:       make(map[string]bool),
		statusCodes:     make(map[int]bool),
		excludeBots:     false,
		ipRanges:        make([]*net.IPNet, 0),
		methods:         make(map[string]bool),
		pathPatterns:    make([]*regexp.Regexp, 0),
		ignoreBlocked:   false,
		blockedStatuses: make(map[int]bool),
	}
}

// ShouldInclude checks if a log entry should be included based on filters
func (f *LogFilter) ShouldInclude(entry *models.LogEntry) bool {
	// Check blocked status
	if f.ignoreBlocked && f.IsBlocked(entry) {
		return false
	}

	// Check bot filter
	if f.excludeBots && entry.IsBot {
		return false
	}

	// Check country filter
	if len(f.countries) > 0 {
		if !f.countries[entry.Country] {
			return false
		}
	}

	// Check status code filter
	if len(f.statusCodes) > 0 {
		if !f.statusCodes[entry.Status] {
			return false
		}
	}

	// Check method filter
	if len(f.methods) > 0 {
		if !f.methods[entry.Method] {
			return false
		}
	}

	// Check IP range filter
	if len(f.ipRanges) > 0 {
		inRange := false
		for _, ipRange := range f.ipRanges {
			if ipRange.Contains(entry.IP) {
				inRange = true
				break
			}
		}
		if !inRange {
			return false
		}
	}

	// Check time range filter
	if f.timeRange != nil {
		if entry.Timestamp.Before(f.timeRange.Start) || entry.Timestamp.After(f.timeRange.End) {
			return false
		}
	}

	// Check path pattern filter
	if len(f.pathPatterns) > 0 {
		matched := false
		for _, pattern := range f.pathPatterns {
			if pattern.MatchString(entry.Path) {
				matched = true
				break
			}
		}
		if !matched {
			return false
		}
	}

	return true
}

// AddCountryFilter adds a country filter
func (f *LogFilter) AddCountryFilter(countries []string) {
	for _, country := range countries {
		f.countries[strings.ToUpper(country)] = true
	}
}

// AddStatusFilter adds a status code filter
func (f *LogFilter) AddStatusFilter(codes []int) {
	for _, code := range codes {
		f.statusCodes[code] = true
	}
}

// AddMethodFilter adds an HTTP method filter
func (f *LogFilter) AddMethodFilter(methods []string) {
	for _, method := range methods {
		f.methods[strings.ToUpper(method)] = true
	}
}

// AddIPRangeFilter adds an IP range filter (CIDR notation)
func (f *LogFilter) AddIPRangeFilter(cidr string) error {
	_, ipNet, err := net.ParseCIDR(cidr)
	if err != nil {
		return err
	}
	f.ipRanges = append(f.ipRanges, ipNet)
	return nil
}

// AddPathPattern adds a path pattern filter (regex)
func (f *LogFilter) AddPathPattern(pattern string) error {
	re, err := regexp.Compile(pattern)
	if err != nil {
		return err
	}
	f.pathPatterns = append(f.pathPatterns, re)
	return nil
}

// SetTimeRange sets the time range filter
func (f *LogFilter) SetTimeRange(start, end time.Time) {
	f.timeRange = &models.TimeRange{
		Start: start,
		End:   end,
	}
}

// SetExcludeBots sets whether to exclude bot traffic
func (f *LogFilter) SetExcludeBots(exclude bool) {
	f.excludeBots = exclude
}

// SetIgnoreBlocked sets whether to ignore blocked requests
func (f *LogFilter) SetIgnoreBlocked(ignore bool) {
	f.ignoreBlocked = ignore
}

// SetBlockedStatusCodes sets the status codes to consider as blocked
func (f *LogFilter) SetBlockedStatusCodes(codes []int) {
	f.blockedStatuses = make(map[int]bool)
	for _, code := range codes {
		f.blockedStatuses[code] = true
	}
}

// IsBlocked checks if a log entry is blocked
func (f *LogFilter) IsBlocked(entry *models.LogEntry) bool {
	if len(f.blockedStatuses) == 0 {
		// Default blocked status codes
		return entry.Status == 403 || entry.Status == 429
	}
	return f.blockedStatuses[entry.Status]
}

// Clear resets all filters
func (f *LogFilter) Clear() {
	f.countries = make(map[string]bool)
	f.statusCodes = make(map[int]bool)
	f.excludeBots = false
	f.ipRanges = make([]*net.IPNet, 0)
	f.timeRange = nil
	f.methods = make(map[string]bool)
	f.pathPatterns = make([]*regexp.Regexp, 0)
	f.ignoreBlocked = false
	f.blockedStatuses = make(map[int]bool)
}

// FilterPresets provides common filter presets

// ErrorsOnly creates a filter for error responses only (4xx, 5xx)
func ErrorsOnly() *LogFilter {
	filter := NewLogFilter()
	for i := 400; i < 600; i++ {
		filter.AddStatusFilter([]int{i})
	}
	return filter
}

// SuccessOnly creates a filter for successful responses only (2xx)
func SuccessOnly() *LogFilter {
	filter := NewLogFilter()
	for i := 200; i < 300; i++ {
		filter.AddStatusFilter([]int{i})
	}
	return filter
}

// NoBots creates a filter that excludes bot traffic
func NoBots() *LogFilter {
	filter := NewLogFilter()
	filter.SetExcludeBots(true)
	return filter
}

// APIOnly creates a filter for API endpoints
func APIOnly() *LogFilter {
	filter := NewLogFilter()
	_ = filter.AddPathPattern(`(?i)/(api|graphql|rest|wp-json)`)
	return filter
}

// RecentActivity creates a filter for recent activity within specified hours
func RecentActivity(hours int) *LogFilter {
	filter := NewLogFilter()
	end := time.Now()
	start := end.Add(-time.Duration(hours) * time.Hour)
	filter.SetTimeRange(start, end)
	return filter
}

// ClientErrorsOnly creates a filter for client errors (4xx)
func ClientErrorsOnly() *LogFilter {
	filter := NewLogFilter()
	for i := 400; i < 500; i++ {
		filter.AddStatusFilter([]int{i})
	}
	return filter
}

// ServerErrorsOnly creates a filter for server errors (5xx)
func ServerErrorsOnly() *LogFilter {
	filter := NewLogFilter()
	for i := 500; i < 600; i++ {
		filter.AddStatusFilter([]int{i})
	}
	return filter
}

// SlowRequests creates a filter for requests slower than threshold (seconds)
func SlowRequests(thresholdSeconds float64) func(*models.LogEntry) bool {
	return func(entry *models.LogEntry) bool {
		return entry.ResponseTime >= thresholdSeconds
	}
}

// LargeResponses creates a filter for responses larger than threshold (bytes)
func LargeResponses(thresholdBytes int64) func(*models.LogEntry) bool {
	return func(entry *models.LogEntry) bool {
		return entry.BytesSent >= thresholdBytes
	}
}

// ByCountries creates a filter for specific countries
func ByCountries(countries []string) *LogFilter {
	filter := NewLogFilter()
	filter.AddCountryFilter(countries)
	return filter
}

// ByStatus creates a filter for specific status codes
func ByStatus(codes []int) *LogFilter {
	filter := NewLogFilter()
	filter.AddStatusFilter(codes)
	return filter
}

// ByMethod creates a filter for specific HTTP methods
func ByMethod(methods []string) *LogFilter {
	filter := NewLogFilter()
	filter.AddMethodFilter(methods)
	return filter
}
